woocommerce/woocommerce-ios
# [Just in Time Messages] JITM banner customisation

Closes: #9570


## Description


In order to show more noticeable messages to our users, we have added the ability to customise the image and badge shown on a Just in Time Message banner.

## Testing instructions


### Test setup
Test setup for this is non-trivial until the JITM message definitions are added - sorry! There are two options:

1. Proxyman/Charles breakpoint and modify the response (easier, but requires intervention for each request): PCYsg-Ffl-p2
2. Set up a WPcom sandbox and ephemeral site with test JITM definitions: P91TBi-a5J-p2

In either case, you'll also need to host the images somewhere, and use the correct conventions when adding them to the JITM definition/JSON response.

#### Using a proxy

Using Charles or Proxyman, set a breakpoint on the following endpoint pattern: `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/?json=true&locale=*&path=/jetpack/v4/jitm*`

Modify the response JSON using the following sample JSON response: 

```
{
  "data": [
    {
      "content": {
        "message": "Accept in-person contactless payments on your iPhone",
        "icon": "",
        "iconPath": null,
        "list": [],
        "description": "Tap to Pay on iPhone is quick, secure, and simple to set up in WooCommerce Payments - no extra terminals or card readers needed.",
        "classes": "",
        "title": "",
        "disclaimer": []
      },
      "CTA": {
        "message": "Set up Tap to Pay",
        "hook": "",
        "newWindow": true,
        "primary": true,
        "link": "https://woocommerce.com/mobile/payments/tap-to-pay"
      },
      "template": "default",
      "ttl": 300,
      "id": "woomobile_tap_to_pay_wcpay_set_up_test_1",
      "feature_class": "woomobile_tap_to_pay_wcpay_set_up",
      "expires": 30,
      "max_dismissal": 50,
      "activate_module": null,
      "is_dismissible": true,
      "is_user_created_by_partner": null,
      "message_expiration": null,
      "assets": {
        "background_image_url": "https://{EPHEMERAL-SITE-HOST}/wp-content/uploads/2023/05/Set-up-Tap-to-Pay-light@3x.png",
        "background_image_dark_url": "https://{EPHEMERAL-SITE-HOST}/wp-content/uploads/2023/05/Set-up-Tap-to-Pay-dark@3x.png",
        "badge_image_url": "https://{EPHEMERAL-SITE-HOST}/wp-content/uploads/2023/05/new-badge@3x.png",
        "badge_image_dark_url": "https://{EPHEMERAL-SITE-HOST}/wp-content/uploads/2023/05/new-badge@3x.png"
      },
      "url": "https://jetpack.com/redirect/?source=jitm-woomobile_tap_to_pay_wcpay_set_up_test_1",
      "jitm_stats_url": "https://pixel.wp.com/b.gif?v=wpcom2&rand=fe304a7d0cabbf0466f5d554a6f9de3a&x_jetpack-jitm=woomobile_tap_to_pay_wcpay_set_up_test_1"
    }
  ]
}
```

#### Using a sandbox

A sample JITM definition (for the sandbox option) is below:

`public_html/wp-content/mu-plugins/0-sandbox.php`
```
<?php
function sandbox_jitm_rules( $rules, $cache ) {
	return array(
		( new \JITM\Message( 'woomobile_tap_to_pay_wcpay_set_up_test_1', 'woomobile_tap_to_pay_wcpay_set_up', $cache ) )
			->show( 'Accept in-person contactless payments on your iPhone',
					'Tap to Pay on iPhone is quick, secure, and simple to set up in WooCommerce Payments - no extra terminals or card readers needed.' )
			->message_path( '/woomobile:my_store:admin_notices/' )
			->with_asset( 'background_image_url', 'https://{EPHEMERAL-SITE-HOST}/wp-content/uploads/2023/05/Set-up-Tap-to-Pay-light@3x.png' )
			->with_asset( 'background_image_dark_url', 'https://{EPHEMERAL-SITE-HOST}/wp-content/uploads/2023/05/Set-up-Tap-to-Pay-dark@3x.png' )
			->with_asset( 'badge_image_url', 'https://{EPHEMERAL-SITE-HOST}/wp-content/uploads/2023/05/new-badge@3x.png' )
			->with_asset( 'badge_image_dark_url', 'https://{EPHEMERAL-SITE-HOST}/wp-content/uploads/2023/05/new-badge@3x.png' )
			->plugin_active( 'woocommerce-payments/woocommerce-payments.php' )
			->with_CTA( 'Set up Tap to Pay',
						'',
						'https://woocommerce.com/mobile/payments/tap-to-pay' )
			->for_sites_of_types( array( 'wpcom', 'jetpack', 'atomic' ) )
			->with_wc_country( 'US' )
			->max_dismissals( 50 )
			->show_again_after( 30 )
			->priority( 150 ),
	);
}
add_filter( 'jetpack_jitm_rules', 'sandbox_jitm_rules', 10, 2 );

```

Image hosting – I uploaded the images to my ephemeral site's media library, and used their URLs in the definition/response. Replace `{EPHEMERAL-SITE-HOST}` with your site's domain if you follow the same route. While it remains up, you're welcome to use `pale-pirate-peach.atomicsites.blog` if you like, just to save a step of hosting the images.


### Testing in the app

Whichever route you choose:

1. Launch the app
2. Observe that the JITM banner is shown
3. Observe that the specified image is shown for your current light/dark mode
4. Switch display mode
5. Observe that the specified images are updated


Checking the cache:

1. Pull to refresh or relaunch the app
2. Observe that cached images are used and load quickly


Set up your image hosting so that the background image is not available/returns an error when requested

1. Pull to refresh
2. Observe that the fallback image is shown


## Screenshots


Note that at the start, there's no badge and a default image is used, then the image is swapped for a remotely loaded one, which changes with the display mode. Finally, the badge is added, and the layout adjusts accordingly.

https://user-images.githubusercontent.com/2472348/236420168-f32a1e46-bc15-48d6-8c7f-4da858ff3f43.mp4



---
